### PR TITLE
feat(executor): Use `TableCatalog` and `StateTable` in `DynamicFilter` (only writes)

### PR DESCRIPTION
**Pull request by jon-chuang**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

Use `TableCatalog` and `StateTable` for `DynamicFilter`

Next 2 PRs:
- State store reads and cache behaviour. Add metrics for managed state (?).
- Unit tests

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
Tracking: https://github.com/singularity-data/risingwave/issues/3419

### CODE REVIEW

File: proto/stream_plan.proto [hunk 1]
```diff
@@ -113,10 +113,9 @@ message HashJoinNode {
 message DynamicFilterNode {
   uint32 left_key = 1;
   // Must be one of <, <=, >, >=
-  expr.ExprNode condition = 3;
+  expr.ExprNode condition = 2;
   // Used for internal table states.
-  // catalog.Table left_table = 4;
-  // uint32 dist_key_l = 8;
+  catalog.Table left_table = 3;
```
Comment (BowenXiao1999) [inline proto/stream_plan.proto:118]: Will there be right_table?

File: src/frontend/src/optimizer/plan_node/stream_dynamic_filter.rs [hunk 2]
```diff
@@ -83,6 +87,37 @@ impl ToStreamProst for StreamDynamicFilter {
...
 }
+
+fn infer_internal_table_catalog(input: PlanRef, left_key_index: usize) -> TableCatalog {
+    let base = input.plan_base();
+    let schema = &base.schema;
+
+    let append_only = input.append_only();
+    let dist_keys = base.dist.dist_column_indices().to_vec();
+
+    // The pk of hash join internal table should be join_key + input_pk.
```
Comment (BowenXiao1999) [inline src/frontend/src/optimizer/plan_node/stream_dynamic_filter.rs:107]: update comment

File: src/stream/src/executor/dynamic_filter.rs [hunk 3]
```diff
@@ -29,27 +30,144 @@ use risingwave_expr::expr::expr_binary_nonnull::new_binary_expr;
...
 use super::barrier_align::*;
-use super::error::StreamExecutorError;
+use super::error::{StreamExecutorError, StreamExecutorResult};
 use super::monitor::StreamingMetrics;
 use super::{BoxedExecutor, BoxedMessageStream, Executor, Message, PkIndices, PkIndicesRef};
 use crate::common::StreamChunkBuilder;
 use crate::executor::PROCESSING_WINDOW_SIZE;
 
-pub struct DynamicFilterExecutor {
+pub struct RangeCache<S: StateStore> {
```
Comment (BowenXiao1999) [inline src/stream/src/executor/dynamic_filter.rs:43]: Now it is not `pub`?

File: src/stream/src/executor/dynamic_filter.rs [hunk 4]
```diff
@@ -29,27 +30,144 @@ use risingwave_expr::expr::expr_binary_nonnull::new_binary_expr;
...
+    #[allow(unused)]
+    num_rows_stored: usize,
+    #[allow(unused)]
+    capacity: usize,
+    current_epoch: u64,
+}
+
+type ScalarRange = (Bound<ScalarImpl>, Bound<ScalarImpl>);
+
+// TODO: add rustdoc comments.
```
Comment (BowenXiao1999) [inline src/stream/src/executor/dynamic_filter.rs:68]: In this PR? 👍

**Review by BowenXiao1999 — COMMENTED** [review]
Will review later

File: proto/stream_plan.proto (hunk 1 above)
Comment (jon-chuang) [reply proto/stream_plan.proto:118]: I think so, it may be a single row/value

File: src/stream/src/executor/dynamic_filter.rs (hunk 3 above)
Comment (jon-chuang) [reply src/stream/src/executor/dynamic_filter.rs:43]: I guess I should move it to managed_state. There's not much in common to group it all there but I guess it's the convention so far.

File: proto/stream_plan.proto (hunk 1 above)
Comment (jon-chuang) [reply proto/stream_plan.proto:118]: Not sure what's PK for simple agg? (Is it empty?). Basically we want a state table with just table id and col id as value prefix

Comment (BowenXiao1999) [reply proto/stream_plan.proto:118]: Yes, we are allowed to insert a Row(vec![]).

**Review by BowenXiao1999 — APPROVED** [review]
LGTM for the state table infer part. 

Cuz the range_cache do not contains the cache policy, and is just a simple wrapper on previous logic (I think it will not be in this PR?), I think it is ok to approve. 

Better create seperate issue for each the TODOs in code and tracking them.

File: proto/stream_plan.proto (hunk 1 above)
Comment (jon-chuang) [reply proto/stream_plan.proto:118]: Ok. I think this is achieved if dist_key and pk_indices are both empty.

**Review by st1page — APPROVED** [review]
LGTM

**Pull request merged**